City-of-Helsinki/helsinki-design-system
# HDS-2065: change default cookie domain

## Description

Cookie consents were stored with domain set as `.hel.fi`. Now for security reasons the domain is by default `<subdomain>.hel.fi`.

Could not just store the cookie to a new domain, because there would be cases where `document.cookie` contains data from both domains. And the order in not guaranteed to be same in all browsers - or servers. 

So old stored cookie should be removed and new stored. A version numbering was added to detect, if the cookie is stored the new way or old. `document.cookie` does not return domain or any other data, so no other way to know.

**The version is stored to a new cookie and cookies require consents, so the new version will ask consents again.**

This PR has no documentation updates, there is another [ticket](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2015) for those.



## Related Issue

Closes [HDS-2065](https://helsinkisolutionoffice.atlassian.net/browse/HDS-2065)

[Demo](https://city-of-helsinki.github.io/hds-demo/hds-2065-cookies/?path=/story/components-cookieconsent--simple-modal-version)

## How Has This Been Tested?

Unit tests, local tests, demo testing.

How to test on the demo site:
- goto [this demo](https://city-of-helsinki.github.io/hds-demo/hds-2065-old-cookie-version/?path=/story/components-cookieconsent--simple-modal-version) with old cookie consent system. 
- it will write cookies the old way.  Note: `github.io` is a public suffix domain. Cookies with domain set to top level (`.github.io`) are not allowed on public suffix sites.
- you can check the stored cookies in the console ("application" tab in Chrome, "cookies" are in the sidebar)
- next goto [demo of this PR](https://city-of-helsinki.github.io/hds-demo/hds-2065-cookies/?path=/story/components-cookieconsent--simple-modal-version). Because of new version cookie, consents are asked again. Before user gives consents, the old ones are converted to new domain and there is now a version cookie. Cookies are also moved to new domain, except that is not possible to test in `github.io`, because all cookie domains must be `subdomain.github.io` (public suffix issue).

You can test cookie fully in localhost, if you add a longer domain for localhost in the [hosts file](https://www.hostinger.com/tutorials/how-to-edit-hosts-file-macos).

Add
`127.0.0.1 local.hds.hel.ninja  `

Then start Storybook and goto http://local.hds.hel.ninja:6006/ and add or modify cookies from the console.

## Add to changelog
- [x] Added needed line to changelog 


[HDS-2065]: https://helsinkisolutionoffice.atlassian.net/browse/HDS-2065?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
